File: BAEKJOON/10775 - 공항/공항.cs
using System;

class Program
{
    // Function to find the root of a gate with path compression
    static int Find(int[] parent, int x)
    {
        if (parent[x] != x)
        {
            parent[x] = Find(parent, parent[x]); // Path compression
        }
        return parent[x];
    }

    // Function to union two gates
    static void Union(int[] parent, int x, int y)
    {
        int rootX = Find(parent, x);
        int rootY = Find(parent, y);
        parent[rootX] = rootY; // Union by setting root of x to root of y
    }

    static void Main()
    {
        int G = int.Parse(Console.ReadLine()); // Read number of gates
        int P = int.Parse(Console.ReadLine()); // Read number of planes

        int[] parent = new int[G + 1];
        
        for (int i = 1; i <= G; ++i)
        {
            parent[i] = i; // Initialize each gate to point to itself
        }

        int dockedPlanes = 0;

        for (int i = 0; i < P; ++i)
        {
            int gi = int.Parse(Console.ReadLine()); // Read the preferred gate for each plane
            
            // Find the highest available gate for docking
            int availableGate = Find(parent, gi);

            if (availableGate == 0)
            {
                break; // If no gate is available, stop processing further planes
            }

            // Dock the plane and update the availability of gates
            Union(parent, availableGate, availableGate - 1);
            dockedPlanes++;
        }

        Console.WriteLine(dockedPlanes); // Output the maximum number of planes docked
    }
}

File: BAEKJOON/10775 - 공항/공항.cpp
#include <iostream>
#include <vector>

using namespace std;

// Function to find the root of a gate with path compression
int find(vector<int>& parent, int x) {
    if (parent[x] != x) {
        parent[x] = find(parent, parent[x]); // Path compression
    }
    return parent[x];
}

// Function to union two gates
void unionGates(vector<int>& parent, int x, int y) {
    int rootX = find(parent, x);
    int rootY = find(parent, y);
    parent[rootX] = rootY; // Union by setting root of x to root of y
}

int main() {
    int G, P;
    cin >> G >> P; // Read number of gates and planes

    vector<int> parent(G + 1);
    for (int i = 1; i <= G; ++i) {
        parent[i] = i; // Initialize each gate to point to itself
    }

    int docked_planes = 0;

    for (int i = 0; i < P; ++i) {
        int gi;
        cin >> gi; // Read the preferred gate for each plane

        // Find the highest available gate for docking
        int available_gate = find(parent, gi);
        
        if (available_gate == 0) {
            break; // If no gate is available, stop processing further planes
        }

        // Dock the plane and update the availability of gates
        unionGates(parent, available_gate, available_gate - 1);
        docked_planes++;
    }

    cout << docked_planes << endl; // Output the maximum number of planes docked

    return 0;
}